feat: Url Shortener

raysubham opened the pull request:
## What does this PR do?

<!-- Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change. -->

Fixes # (issue)
https://github.com/formbricks/formbricks/issues/880

Loom Video: https://www.loom.com/share/0a89c1bbcf2c49d5ac63232be6a06fbd?sid=3a4f29bc-0fce-42ab-9fc6-ee7a52275971
 
## Type of change

<!-- Please mark the relevant points by using [x] -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Chore (refactoring code, technical debt, workflow improvements)
- [x] Enhancement (small improvements)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change adds a new database migration
- [x] This change requires a documentation update

## How should this be tested?

<!-- Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration -->

- Test A
- Test B

## Checklist

<!-- We're starting to get more and more contributions. Please help us making this efficient for all of us and go through this checklist. Please tick off what you did  -->

- [x] Added a screen recording or screenshots to this PR
- [ ] Filled out the "How to test" section in this PR
- [x] Read the [contributing guide](https://github.com/formbricks/formbricks/blob/main/CONTRIBUTING.md)
- [x] Self-reviewed my own code
- [x] Commented on my code in hard-to-understand bits
- [x] Ran `pnpm build`
- [x] Checked for warnings, there are none
- [x] Removed all `console.logs`
- [x] Merged the latest changes from main onto my branch with `git pull origin main`
- [x] My changes don't cause any responsiveness issues
- [ ] Updated the Formbricks Docs if changes were necessary

Dhruwang commented on apps/web/app/(app)/environments/[environmentId]/Navigation.tsx:
```diff
@@ -185,6 +188,15 @@ export default function Navigation({
+          icon: LinkIcon,
+          label: "Link Shortener",
+          href: pathname,
+          hidden: widgetSetupCompleted,
```
Please remove `hidden: widgetSetupCompleted` as it will hide "Link Shortener" if a users has completed widget setup 😊

Dhruwang commented on apps/web/app/i/[shortUrl]/page.tsx:
```diff
@@ -0,0 +1,11 @@
+  const fullUrl = await getFullUrl(params.shortUrl);
+
+  if (!fullUrl) redirect("/");
+  if (fullUrl) redirect(fullUrl);
```
On line 7, we are alreday checking whether fullUrl exist, so we can just avoid rechecking it on next line

```
if (fullUrl) redirect(fullUrl);
  else {
    redirect("/");
  } 
```

Dhruwang commented on packages/lib/services/urlshortener.ts:
```diff
@@ -0,0 +1,84 @@
+import { DatabaseError } from "@formbricks/types/v1/errors";
+
+// Create short url and return it
+export const createShortUrl = async (fullUrl: string) => {
```
Please add return type for this method 😊

Dhruwang commented on packages/lib/services/urlshortener.ts:
```diff
@@ -0,0 +1,84 @@
+};
+
+// Get full url from short url and increment the hits count
+export const getFullUrl = async (shortUrlParam: string) => {
```
return type :)

Dhruwang commented:
Thank you @raysubham for the PR, looks pretty great 💪😊

I left some comments for improving code quality
Also I noticed one minor issue, where after closing url shortening modal, its fields does not gets reset


https://github.com/formbricks/formbricks/assets/67850763/70e5cb05-2fd8-484d-93f5-e757ce0a14a1

Dhruwang commented on apps/web/env.mjs:
```diff
@@ -70,6 +71,7 @@ export const env = createEnv({
     NEXT_PUBLIC_POSTHOG_API_KEY: z.string().optional(),
     NEXT_PUBLIC_POSTHOG_API_HOST: z.string().optional(),
     NEXT_PUBLIC_SENTRY_DSN: z.string().optional(),
+    NEXT_PUBLIC_SURVEY_BASE_URL: z.string().optional(),
```
Please use the existing serverside env variables (`SURVEY_BASE_URL`) and pass them through to the client component. We can’t use NEXT_PUBLIC_ env variables because we need all env variables to be evaluated at runtime to support the docker deployment.

raysubham commented:
@Dhruwang Thanks so much for catching the issues in the code. I have addressed all the comments & bugs. Feel free to re-test this once when you have time 😊

Dhruwang commented on packages/database/schema.prisma:
```diff
@@ -500,3 +500,12 @@ model User {
+  updatedAt DateTime @updatedAt @map(name: "updated_at")
+  fullUrl   String
+  shortUrl  String
+  hits      Int      @default(0)
```
Can you please remove hits and corresponding service logic, as of now we don't need that count 😊

raysubham replied:
Sure, I removed the `hits` from prisma schema.

Dhruwang commented:
Thank you for addressing all the updates! 🙌🏻
Just one minor adjustment left to make.
